Add new `FiniteCycle` type as the return type of `Collection.cycled(times:)`

LemonSpike opened the pull request:
<!--
    Thanks for contributing to Swift Algorithms!

    If this pull request adds new API, please add '?template=new.md'
    to the URL to switch to the appropriate template.

    Before you submit your request, please replace the paragraph
    below with the relevant details, and complete the steps in the
    checklist by placing an 'x' in each box:
    
    - [x] I've completed this task
    - [ ] This task isn't completed
-->

This PR aims to fix #99. I have also split up some relevant tests into separate methods for empty collections and changed some tests for `FiniteCycle`. I have also added tuple labels to `Product2` when using the `Iterator.next()` method and also in the `subscript` method. This is a cosmetic improvement, but I felt it may read better.

### Checklist
- [x] I've added at least one test that validates that my change is working, if appropriate
- [x] I've followed the code style of the rest of the project
- [x] I've read the [Contribution Guidelines](../CONTRIBUTING.md)
- [x] I've updated the documentation if necessary

xwu commented on Sources/Algorithms/Cycle.swift:
```diff
@@ -56,6 +56,47 @@ extension Cycle: Sequence {
 extension Cycle: LazySequenceProtocol where Base: LazySequenceProtocol {}
 
+
+/// A collection wrapper that repeats the elements of a base collection for a finite number of times.
```
Nit: Please wrap to 80 columns, here and elsewhere.

xwu commented on Sources/Algorithms/Cycle.swift:
```diff
@@ -56,6 +56,47 @@ extension Cycle: Sequence {
+/// A collection wrapper that repeats the elements of a base collection for a finite number of times.
+public struct FiniteCycle<Base: Collection> {
+  /// The collection to repeat.
+  public let base: Base
```
Let's make this `internal`, as well as `times`.

This is because equality (we could conform this type to `Equatable`, but it doesn't have to be here in this PR) should be based on elementwise comparisons, and `[1, 2, 3].cycled(times: 6)` should be equivalent to `[1, 2, 3, 1, 2, 3].cycled(times: 3)`. Having `base` and `times` publicly accessible would raise the question of whether these properties are "salient" ones for the purposes of considering equivalence.

LemonSpike replied:
Should I also change `Cycle.base` to internal as well? This was why I used public here.

xwu reviewed (commented):
👍

timvermeulen commented:
Thanks for taking this on @LemonSpike, this is a great start! Your `FiniteCycle` correctly delegates all of the work to an underlying `Product2<Range<Int>, Base>`. However, the real benefit will come from being able to efficiently implement all of the `Collection` methods that `Product2` already implements. By forwarding all public `Collection` and `BidirectionalCollection` methods to an underlying `Product2` collection, `FiniteCycle` will have all the same performance characteristics as `Product2`. Want to have a go at that?

LemonSpike commented:
Thanks for the feedback @timvermeulen! I have tried to add the `Collection` protocol conformances this time by forwarding to an underlying `Product2` collection. Are these what you had in mind? Thanks.

timvermeulen commented:
This is indeed what I had in mind, great work! The `Collection` conformance looks good, it implements all the relevant methods. There are just a few things left to do:
- We can get rid of the `Sequence` conformance entirely — normally it's needed in case the base type is only a `Sequence` and `Collection` cannot be implemented, but in the case of `FiniteCycle` we already require that `Base: Collection`, so it's no longer needed. In the same vein, the type needs `LazyCollectionProtocol` conformance rather than `LazySequenceProtocol`.
- `FiniteCycle` is still missing some conformances that `Product2` does have, implemented [here](https://github.com/apple/swift-algorithms/blob/main/Sources/Algorithms/Product.swift#L433-L435).
- Most public methods are missing `@inlinable` annotations.
- The `Cycle.md` guide needs a slight tweak to reflect that `cycled(times:)` now has a new return type.

LemonSpike commented:
Hi @timvermeulen, thanks for the feedback. Re: your points, I tried to implement your suggestions as follows:

1. I removed the `Sequence` conformance entirely. To conform to `LazyCollectionProtocol` I had to conform to `LazySequenceProtocol` and `LazyCollectionProtocol` together where `Base: LazyCollectionProtocol`, because I couldn't inherit `LazySequenceProtocol` conformance automatically.
 2. Managed to add these as suggested, but because the `FiniteCycle.Index` is a typealias for `Product2.Index`, the `Index` `Hashable` conformance was already synthesized.
 3. Added `@inlinable` to all public methods, by making the `let product` `public`.
 4. Markdown file edited but reworded without `despite its name` since the name no longer includes `Sequence`.

Let me know your thoughts. Thanks.

timvermeulen commented:
Apologies for the delay, @LemonSpike!

> 1. I removed the `Sequence` conformance entirely. To conform to `LazyCollectionProtocol` I had to conform to `LazySequenceProtocol` and `LazyCollectionProtocol` together where `Base: LazyCollectionProtocol`, because I couldn't inherit `LazySequenceProtocol` conformance automatically.

We can get rid of the `Iterator` type and `makeIterator` method entirely. The typechecker did get confused about what the `Element` type is after I tried doing this, so you might need to add `typealias Element = Base.Element` to the `Collection` conformance.

> 2. Managed to add these as suggested, but because the `FiniteCycle.Index` is a typealias for `Product2.Index`, the `Index` `Hashable` conformance was already synthesized.

Great point! Let's for now give `FiniteCycle` its own `Index` type that wraps a `Product2.Index`, just so in the future we have the freedom to change the internals if the need ever arises.

> 3. Added `@inlinable` to all public methods, by making the `let product` `public`.

While we make all public and internal methods `@inlinable`, for stored properties `@usableFromInline` suffices, so there's no need to make the property public.

> 4. Markdown file edited but reworded without `despite its name` since the name no longer includes `Sequence`.

That looks good, could you also have it mention the conditional `LazyCollectionProtocol` conformance, similar to the `Cycle` description? And there's a note a bit higher up about `cycled(times:)` being a combination of two existing standard library functions which we can probably get rid of altogether.

> 5. I changed some `@usableFromInline` `internal` methods to `@inlinable` because of #109.

💯

timvermeulen commented on Sources/Algorithms/Product.swift:
```diff
@@ -44,7 +44,8 @@ extension Product2: Sequence {
     
     @inlinable
-    public mutating func next() -> (Base1.Element, Base2.Element)? {
+    public mutating func next() -> (element1: Base1.Element,
+                                    element2: Base2.Element)? {
```
Let's keep this tuple unlabeled — `Zip2Sequence`'s element tuple doesn't have labels either, and generic labels such as `element1` and `element2` don't really add any clarity over `.0` and `.1`.

LemonSpike replied:
I am happy to do this, but I still think that the labels improve readability for users who don't know about tuple indexing. I saw we use more descriptive names for public APIs and type names, so I think it could help users understand the API. Let me know your thoughts.

timvermeulen replied:
These labels don't improve the users' understanding, because they don't convey any useful information. It's impossible for us to give the tuple elements descriptive names because at this level of abstraction we have nothing to go on. To quote the [API Design Guidelines](https://swift.org/documentation/api-design-guidelines/#naming):

>**Omit needless words.** Every word in a name should convey salient information at the use site.

Feel free to start a discussion about this topic on the forums, as this is equally relevant to the `Zip2Sequence` element type. For the time being we'd like to stick with the precedent set by the standard library.

natecook1000 commented:
@swift-ci Please test

LemonSpike commented:
@swift-ci Please test

natecook1000 commented:
@swift-ci Please test

natecook1000 commented:
Looks great, @LemonSpike — thanks for this addition! 🎉
